### DISCUSSION TITLE
ManyToOne cascade cannot be disabled if the field is nullable

### DISCUSSION THREAD

**Discussion by araines** [mikro-orm/mikro-orm#3312] [Q&A] [answered]
**Describe the bug**
The desired SQL schema I'm aiming for is like this:

```
CREATE TABLE deployment (
    id              int(11) unsigned    NOT NULL    AUTO_INCREMENT,
    PRIMARY KEY (id)
);

CREATE TABLE commit (
    id              int(11) unsigned    NOT NULL    AUTO_INCREMENT,
    deployment_id   int(11) unsigned                DEFAULT NULL,
        
    PRIMARY KEY (id),
    FOREIGN KEY (deployment_id) REFERENCES deployment(id)
);
```

ie. I should not be able to delete a deployment if it contains commits, but a commit is only optionally related to a deployment.

I have my Commit entity configured as so:
```
@Entity()
export default class Commit {
  @PrimaryKey()
  id: number

  @ManyToOne('Deployment', { nullable: true, cascade: [] })
  deployment?: Deployment
}
```

But running the schema CLI it keeps wanting to put an `on delete set null` statement in:

```
alter table `commit` add constraint `commit_deployment_id_foreign` foreign key (`deployment_id`) references `deployment` (`id`) on delete set null;
```

It is possible I'm misunderstanding how this is supposed to work, I've only just picked up MikroORM!

**Versions**

| Dependency | Version |
| - | - |
| node | 17.5.0 |
| typescript | 4.7.4 |
| mikro-orm | 5.2.3 |
| your-driver | mysql |

**Comment by B4nan** [comment, marked as answer]
The `cascade` option is for application level cascading, that's not the same as referential integrity - it actually works the other way: 
- what you set in your entity definition means there won't be any action cascaded to that relation if you remove the owning entity
- what you want is to control what the owning entity will do if that relation is removed (so the opposite direction) 

You can use `onDelete` to control the referential integrity directly. Also by doing `cascade: []` you also disable cascade persist, I doubt you want that (especially if you don't know what that means 😉).

```ts
@ManyToOne('Deployment', { nullable: true, onDelete: 'restrict' })
deployment?: Deployment
```

(default integrity for a FK is `restrict` or `no action` - those are aliases)

**Reply by araines** [reply]
Ah amazing, thanks for the quick and detailed explanation.  I'd never come across the `restrict` option before, almost certainly because its the default behaviour so had never needed to set it explicitly.  I definitely don't want to disable cascade persist!

**Reply by araines** [reply]
Actually @B4nan small follow up question - given the default behaviour is `restrict` and that is what the referential integrity is set to normally, why does the behaviour change when there is a nullable field on a ManyToOne?

**Reply by B4nan** [reply]
I believe people mostly want that when they have a nullable M:1 or 1:1 relation, there is some logic around the defaults. So nothing else than me picking the way it works :D
